### PR TITLE
Port the completer to Hy

### PR DESCRIPTION
**Pull request by vodik**
Deals with the mangling problems in the REPL by properly mangling and unmangling before completing.

Also ported it to Hy, since, well, might as well be as self-hosting as possible.

### CODE REVIEW

File: hy/lex/parser.py [hunk 1]
```diff
@@ -44,7 +44,8 @@ def unicode_char_to_hex(uchr):
         # Covert a unicode char to hex string, without prefix
         return uchr.encode('unicode-escape').decode('utf-8').lstrip('\\U').lstrip('\\u').lstrip('0')
 
-    assert s
+    if not s:
+        return ""
```
Comment (Kodiologist) [inline hy/lex/parser.py]: Is this line (and the corresponding one in `unmangle`) returning the same type as the rest of the function?

File: hy/completer.hy [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+      (import readline)
+      (except [ImportError]
+        (setv docomplete False)))))
+
+(defclass Completer []
+  (defn --init-- [self &optional [namespace {}]]
+    (when (not (instance? dict namespace))
+      (raise (TypeError "namespace must be a dictionary")))
+
+    (setv self.namespace namespace
```
Comment (Kodiologist) [inline hy/completer.hy]: Shouldn't you copy `namespace` (or set the default argument of `namespace` to `None` and then assign `{}` in the body)?

**Comment by Kodiologist** [comment]
I think you should allow completion of the mangled form, too. Thus, if you imported a symbol `is_foo` from a Python module, `is_f` will be recognized by the completer, as a user would probably expect.

File: hy/completer.hy [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+                               (instance? string-types key))
+                     (viewkeys builtins.--dict--)
+                     (viewkeys (get hy.macros.-hy-macros None))
+                     (viewkeys namespace)]
+          self.tag-path [(viewkeys (get hy.macros.-hy-tag None))])
+
+    (setv module-name (.get namespace "__name__"))
+    (when module-name
+      (->> (get hy.macros.-hy-macros module-name) (viewkeys) (.append self.path))
+      (->> (get hy.macros.-hy-tag module-name) (viewkeys) (.append self.tag-path))))
```
Comment (Kodiologist) [inline hy/completer.hy]: Why use `->>` and write these lines backwards when you could just write them forwards?

File: hy/completer.hy [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+
+  (defn attr-matches [self text]
+    (setv match (re.match r"(\S+(\.[\w-]+)*)\.([\w-]*)$" text))
```
Comment (Kodiologist) [inline hy/completer.hy]: This doesn't seem to allow for attributes with weird characters (that would need to be mangled) in their name.

File: hy/lex/parser.py (hunk 1 above)
Comment (vodik) [reply hy/lex/parser.py]: Yeah, so I don't mind putting an error back in (probably should raise a `ValueError` though) and handling the edge case inside the completer.

Comment (Kodiologist) [reply hy/lex/parser.py]: That sounds reasonable to me.

File: hy/completer.hy (hunk 2 above)
Comment (vodik) [reply hy/completer.hy]: Yeah, good catch. I ported over the old code without thinking hard enough.

**Comment by ekaschalk** [comment]
Have you all seen: https://github.com/ekaschalk/jedhy? Example usages here: https://github.com/ekaschalk/jedhy/blob/master/tests/test_models.hy

It performs completion on all constructs (macros, compile table, module members, etc...), among other actions.

I use it in the develop branch of `hy-mode`, pending merge while I'm still learning how to manage its auto-installation.

**Comment by Kodiologist** [comment]
That's cool. Perhaps the completion parts should be moved into Hy, or perhaps its own completion should be replaced with Hy's.

**Comment by vodik** [comment]
@ekaschalk I'd be in favour of it, and we always leverage optional dependencies for it. The Python repl doesn't  have any autocomplete and we can always keep a minimal install by default. Do something like `pip install hy[autocomplete]` or something.

**Comment by Kodiologist** [comment]
I'm closing this for lack of response.

**Pull request closed without merging**